### ISSUE TITLE
Leeü 

### ISSUE THREAD

**Issue by W1zzardTPU** [upx/upx#104]
## What's the problem (or question)?
Is there any way to keep the debug directory in the compressed exe, so I can load crashdump files with PDB info?

## What should have happened?
Add command line option to not strip debug directory

## How can we reproduce the issue?
Win32, pack any exe, dumpbin shows missing debug directory

## Please tell us details about your environment.
Win32, UPX 3.94

**Comment by AraHaan** [comment]
Happens on 64 bit executables too.

**Comment by W1zzardTPU** [comment]
Any update on this?

**Comment by AraHaan** [comment]
Looks like there is no update yet on this.

**Comment by W1zzardTPU** [comment]
Yeah surprising, should be a trivial fix for the developers that know how their code works.

**Comment by jreiser** [comment]
A reproducible test case is **required** in order to be specific because of possible dependencies on compiler, linker, etc. Upload ("Attach files by ...") a small PE32 file that contains a Debug directory and is compressible.  Hint:  something like a non-optimized compilation of
```
int x[10000] = {1,2,3};  /* compressible *?
int main(int argc, char *argv[])
{
    int *p = (int *)0;
    return *p;  /* Deliberate error to exercise debugging functionality. */
}
```
Apply gzip or zip to reduce the size of the upload.

**Comment by W1zzardTPU** [comment]
Included source and a 32-bit and 64-bit executable of the Visual Studio default Win32 console project.

[UPXTest.zip](https://github.com/upx/upx/files/9478424/UPXTest.zip)

https://manalyzer.org/report/98e3c6ae16dd9fe7547b697d7158b303

**Comment by jreiser** [comment]
@W1zzardTPU  At run time of the target program, when it comes time to access the data in the DEBUG section, is it accessed using only values that are in the original .exe image file? Or is it accessed using data that is in the memory image of the process?  Or some mixture: some from the .exe, some from RAM?

**Comment by W1zzardTPU** [comment]
The target program never accesses the debug directory by itself. 

The Visual Studio debugger uses the data in that directory to locate the PDB symbol info. If the file listed in the debug directory does not exist, it will offer to browse for the right file. 

If there's no debug directory you simply cannot load the symbols at all, no browse, no anything. It seems the debug directory stores some kind of checksum, which is used by VS to determine if the file provided is the "right" one.

**Comment by jreiser** [comment]
`debugger uses the data in that [debug] directory`: Does the VS debugger look at the .exe file, or at the contents of RAM?

**Comment by W1zzardTPU** [comment]
Great question, I have no idea, also not sure how to test?

**Comment by markus-oberhumer** [comment]
Closing as WONTFIX for now.

**Comment by W1zzardTPU** [comment]
Considering I spend like one minute to fix up the exe each time I do a post-mortem debug on a dmp file, and I will do so for many many years, is there any way to get this feature through a paid sponsorship?

**Comment by markus-oberhumer** [comment]
Well, maybe somebody could tackle this, but I'd guess some more detailed info about the actual problem would help.

**Comment by W1zzardTPU** [comment]
My application is downloaded roughly 1 million times each month, it's UPX compressed.

When there's an application crash, the user can optionally submit the crash dump for analysis.

I load the crash dump into Visual Studio and it immediately complains that it can't find the PDB file (because the debug directory is missing in the UPX EXE). A crashdump from the uncompressed EXE instantly loads the PDB, which gives me a proper stack trace, with link to sources, so I can figure out what caused the crash.

I would expect that this is a common workflow for all shipped applications, so I'm surprised that I'm the only one reporting this issue.

**Comment by jreiser** [comment]
`I spend like one minute to fix up the exe` What **exactly** do you do to "fix" it?  Does the "fixed" file (re-)execute normally, or does it just "debug" correctly?  If it takes one minute to fix "by hand", then could a Python script do it in several seconds?

For instance, do you just choose the descriptor for some PE Section in the compressed file, overwrite that descriptor with the descriptor for a Debug section, then append the contents of the Debug section (taken from the not-compressed file) onto the end of the compressed file?    That is scriptable.  Why cannot the debugger be invoked on the not-compressed file in the first place?  (On *nix the debugger is invoked with two files: the executable, and the crash dump.) Or why cannot the "tail" of the crash dump (the portion with "updated" data values) be overwritten onto a copy of the not-compressed file?

**Comment by W1zzardTPU** [comment]
I use a script to change ImageNtHeaders.OptionalHeader.SizeOfImage in the dump file, so that I can force load the PDB (for manual load VS only checks size). That's 3 clicks, plus a file drag drop, then I have to perform like 6 clicks in the VS GUI to make everything load correctly

> Why cannot 

Typical Microsoft user experience I guess ;)

**Comment by rplociennik** [comment]
What's happening here is, I believe, UPX (unintentionally?) removes the debug "directories" (addressed by `IMAGE_DIRECTORY_ENTRY_DEBUG` as part of the optional headers: https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-image_optional_header32) alongside any data referenced by it (stored elsewhere in the binary).

This data is an array of `IMAGE_DEBUG_DIRECTORY`: https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-image_debug_directory and it's the `IMAGE_DEBUG_TYPE_CODEVIEW` debug information that stores the path to the `.pdb` file as well as crucial identifiers allowing pairing a given executable with its debug symbols.

**Comment by ml1050** [comment]
If your use case only needs that the ORIGINAL debug directory from the uncompressed program should not be zeroed when your program is running, then I think this could be done with little effort.

See:

https://github.com/upx/upx/blob/261609f706c0d57080eccd040606ada2dd83df08/src/pefile.cpp#L2090
https://github.com/upx/upx/blob/261609f706c0d57080eccd040606ada2dd83df08/src/pefile.cpp#L2496

But if the debug directory is needed to be kept uncompressed in the compressed file, then it would be more work.

**Comment by rplociennik** [comment]
I'm not entirely sure if it's needed outside runtime. This data can be easily retrieved offline, so chances are these directories and the data they point to might have to be preserved and kept uncompressed.

I might get a chance to play around with this some more.

Thanks for the insight!
